javaslang/src/main/java/javaslang/concurrent/Future.java

```diff
@@ -802,16 +818,35 @@ default boolean isFailure() {
      * @return A new Future that returns both Future results.
      * @throws NullPointerException if {@code that} is null
      */
-    @SuppressWarnings("unchecked")
     default <U> Future<Tuple2<T, U>> zip(Future<? extends U> that) {
         Objects.requireNonNull(that, "that is null");
-        final Promise<Tuple2<T, U>> promise = Promise.make(executorService());
+        return zipWith(that, Tuple::of);
+    }
+
+    /**
+     * Returns a this and that Future result combined using a given combinator function.
+     * <p>
+     * If this Future failed the result contains this failure. Otherwise the result contains that failure or
+     * a combination of both successful Future results.
+     *
+     * @param that Another Future
+     * @param combinator The combinator function
+     * @param <U>  Result type of {@code that}
+     * @param <R>  Result type of {@code f}
+     * @return A new Future that returns both Future results.
+     * @throws NullPointerException if {@code that} is null
+     */
+    @SuppressWarnings("unchecked")
+    default <U, R> Future<R> zipWith(Future<? extends U> that, Function2<T, U, R> combinator) {
```

Comment:
Please use a `BiFunction` instead of a `Function2` (which is a `BiFunction`) and add upper/lower type bounds:

```java
<U, R> Future<R> zipWith(Future<? extends U> that, BiFunction<? super T, ? super U, ? extends R> combinator)
```